test/versioned/fastify/fastify-3-naming.tap.js

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2021 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+const tap = require('tap')
+const requestClient = require('request')
+const helper = require('../../lib/agent_helper')
+
+let callCount = 0
+const loadMiddleware = async (fastify) => {
+  function testMiddleware(req, res, next) {
+    callCount++
+    next()
+  }
+
+  await fastify.register(require('middie'))
+
+  fastify.use(testMiddleware)
+}
+
+/**
+ * Single function to register all the routes used by the test
+ *
+ * @todo Would this be better closer to test, or is it good here
+ */
+const configureFastifyServer = async (fastify) => {
+  /**
+   * Route's callback is an async function, and response is returned
+   */
+  fastify.get('/async-return', async () => {
+    return { called: '/async-return' }
+  })
+
+  /**
+   * Route's callback is an async function, and response is sent via reply
+   */
+  fastify.get('/async-reply-send', async (request, reply) => {
+    reply.send({ called: '/async-reply-send' })
+  })
+
+  /**
+   * Route's callback is not an async function, and response is sent via reply
+   */
+  fastify.get('/sync-reply-send', (request, reply) => {
+    reply.send({ called: '/sync-reply-send' })
+  })
+
+  /**
+   * Register a route via plugin to make sure our wrapper catches these
+   */
+  fastify.register(function (fastifyInstance, options, done) {
+    fastifyInstance.get('/plugin-registered', async () => {
+      return { called: '/plugin-registered' }
+    })
+    done()
+  }, {})
+
+  await loadMiddleware(fastify)
+}
+
+let testCount = 0
+const testUri = (uri, agent, test, port) => {
+  agent.on('transactionFinished', (transaction) => {
+    testCount++
+    test.equals(
+      `WebFrameworkUri/Fastify/GET/${uri}`,
+      transaction.getName(),
+      `transaction name matched for ${uri}`
+    )
+  })
+
+  requestClient.get(`http://127.0.0.1:${port}${uri}`, function (error, response, body) {
+    const result = (body = JSON.parse(body))
+    test.equals(result.called, uri, `${uri} url did not error`)
+  })
+}
+
+const routesToTest = [
+  '/async-return',
+  '/async-reply-send',
+  '/sync-reply-send',
+  '/plugin-registered'
+]
+
+tap.test('Test Transaction Naming - Standard Export', (test) => {
+  test.autoend()
+
+  test.beforeEach(async (t) => {
+    const agent = helper.instrumentMockedAgent({
+      feature_flag: {
+        fastify_instrumentation: true
+      }
+    })
+    const fastify = require('fastify')()
+    await configureFastifyServer(fastify)
+
+    t.context.agent = agent
+    t.context.fastify = fastify
+  })
+
+  test.afterEach((t) => {
+    const { agent, fastify } = t.context
+
+    helper.unloadAgent(agent)
+    fastify.close()
+  })
+
+  createTests(test)
+})
+
+tap.test('Test Transaction Naming - Fastify Property', (test) => {
```

Comment:
We can prob further extract these so we pass a setup function in and then generate test suites, although we will want to make sure things don't get too abstract. Took the simpler approach for now as I don't know what sorts of tests/structures we'll end up with.